### ISSUE TITLE
Potential memory leak

### ISSUE THREAD

**Issue by pecnik** [hmans/miniplex#92]
When an entity gets destroyed, its **world.entities** slot is not being reused when a new entity is created.

This might a problem for shooter games where a lot of
projectile entities are frequently created and destroyed,
because the world.entries array will grow indefinitely.

Example code to illustrate what I mean:
```
const world = new World<Entity>();

for (let i = 0; i < 10; i++) {
    const entity = world.createEntity({
        /* ... */
    });

    if (i % 2 === 0) {
        world.destroyEntity(entity);
    }
}

console.log(world.entities); // [ null, Entity, null, Entity, null, Entity, null, Entity, ... ]
```
---

It would be nice if it could be have something like this:
```
const e1 = world.createEntity({}); // e1.__miniplex.id = 0
const e2 = world.createEntity({}); // e2.__miniplex.id = 1
const e3 = world.createEntity({}); // e3.__miniplex.id = 3

// Free up the world.entities[1] slot
world.destroyEntity(e2); 
console.log(world.entities); // [e1, null, e3]

const e4 = world.createEntity({}); // e4.__miniplex.id = 2 (uses the first free slot)
console.log(world.entities); // [e1, e4, e3]
```

---

_P.S. 

I've wanted a library exactly like this one for years.
Tried to write my own multiple times,
with varying degrees of success.

And im so glad someone managed to write a nice, intuitive, typesafe ecs library.

Thank you_

**Comment by hmans** [comment]
Hi, thanks for submitting the issue, and bonus thanks for submitting a PR to go along with it! I need to consider your solution for a bit, as I am not terribly keen on re-using formerly issued IDs.

The `entities` array currently serves two concerns:

- Tracking all entities that are currently "alive" in the world
- Issuing IDs to newly created entities

I might split up these two concerns and change `entities` to be mutating (like it used to be in an earlier version), and generate IDs using a separate mechanism.

Either way, I will spend some more time with Miniplex soon to give it some upgrades, and will put this issue on my list of things to tackle.

**Comment by pecnik** [comment]
Great, can't wait to see the upgrades!

---

After some thought I think my current PR won't do.

When an entity is destroyed and a new one re-uses the ID,
another entity might be referencing that ID ( like a followEntityComponent for example ).

The current implementation does not have this issue.

Maybe a generational-index would be a more proper solution?

**Comment by hmans** [comment]
> another entity might be referencing that ID

Yup, that is the very thing I'm worried about. But I will make sure to tackle this issue when I do the next round of updates to this library.

**Comment by hamzakubba** [comment]
Hey, there's a O(1) solution to this... rather than explain it and give credit, I can do both with this link 🙂:
https://youtu.be/oewDaISQpw0?t=658

**Comment by hamzakubba** [comment]
Update on this after our chats, putting the summary here for potential future work.

You pointed out a problem with the array-element-swap-deletion method, which is that changing the order of array elements during a loop would break things, so the O(1) solution above would not work for immediate (i.e. non-queued) deletion.

So some options to consider:
1. There can be an option when creating an ECS world to enable memory-efficient deletions at all, the trade-off being that entity order can change at flush time.
2. Instead of a world-creation option, this can instead be an option passed to the delete function, though I feel like that may hide the behavior. Plus if you do it in some places, then you might as well do it everywhere, as a developer, so I don't see why someone would want that.
3. If memory-efficient deletion is enabled, then deleting an entity immediately replaces that entity with a null (as it does now), and additionally tracks the indices of all the deleted entities in the world entities array as well as each archetype's array. Then, at flush time, all the nulls can be swapped to the end and popped, using the same technique, to achieve memory efficiency at a much cheaper computational cost than splicing.

**Comment by hmans** [comment]
Hi, I'm preparing the next minor version of miniplex, and as it's already containing some (very minor) breaking changes, I'm using the opportunity to tackle this thing here.

After giving it a lot of thought (months of it, quite literally, but unintendedly), I think what I would like to do is to actually _disconnect_ the numerical ID (which currently only exists in the internal `__miniplex` component) from the entity's position in the `entities` array, and use the shuffle-delete method (see the video that Hamza posted) to keep it compact. It was a nice property to have, but I think the cost currently outweighs its value, especially considering that Miniplex puts such a big focus on using object identities instead of ID values for referencing things, and a proper ID mechanism should ideally be implemented in userland.

Having said that, I think I'd like to make the following changes, and would appreciate a quick sanity check:

- When destroying an entity, I will immediately fill its free slot with the entity that's currently at the tail of the list, and shorten it by one.
- I will change the ID generation mechanism to use a simple incrementing number to assign new IDs (instead of using `entities.length`).

On top of that, I think I would enjoy also doing the following, please voice your thoughts:

- I might make `world.entities` private. The user should really _not_ use it for any userland logic, and its structure should remain an implementation detail.

Any feedback is highly appreciated.

**Comment by hmans** [comment]
Fix coming in 0.11, by way of #131.
